### PR TITLE
Add more IntelliSense features

### PR DESCRIPTION
**Pull request by LazarusX**
Add more IntelliSense features:
- Image URL validation
- Dockerfile path validation
- Hover preview of Dockerfile content on image URL
- Go-to-Dockerfile on image URL

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -21,9 +21,22 @@ export function activate(context: vscode.ExtensionContext) {
 
     Utility.registerDebugTelemetryListener();
 
-    context.subscriptions.push(vscode.languages.registerCompletionItemProvider([{ language: "json" }, { language: "jsonc" }], new JsonCompletionItemProvider(), "\"", ".", ":"));
-
-    const outputChannel: vscode.OutputChannel = vscode.window.createOutputChannel(Constants.edgeChannel);
+    const configIntelliSenseProvider = new ConfigIntelliSenseProvider();
```
Comment (adashen) [inline src/extension.ts]: Is there any reason we put three kinds of provider into one class?

File: src/common/configIntelliSenseProvider.ts [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        if (this.locationMatch(location, Constants.imgDeploymentManifestJsonPath)) {
+            const moduleToImageMap: Map<string, string> = new Map();
+            const imageToDockerfileMap: Map<string, string> = new Map();
+
+            try {
+                await Utility.setSlnModulesMap(path.dirname(document.uri.fsPath), moduleToImageMap, imageToDockerfileMap);
+
+                const node: parser.Node = location.previousNode;
+                const imageUrl: string = node.value;
+                const imageUrlUnwrapped: string = imageUrl.slice(2, -1); // remove the wrapping "${" and "}"
```
Comment (adashen) [inline src/common/configIntelliSenseProvider.ts]: edge case. If the imageUrl is not start with ${ and end with }, And after slicing imageUrl.slice(2, -1), the value could be got from the map.

File: src/common/configIntelliSenseProvider.ts [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+            } catch {
+                return undefined;
+            }
+        }
+
+        return undefined;
+    }
+
+    public async provideDefinition(document: vscode.TextDocument, position: vscode.Position): Promise<vscode.Definition> {
+        const location: parser.Location = parser.getLocation(document.getText(), document.offsetAt(position));
```
Comment (adashen) [inline src/common/configIntelliSenseProvider.ts]: duplicate code between provideDefinition and provideHover, extract an utility instead?

File: src/common/configIntelliSenseProvider.ts [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+                const moduleName: string = moduleNode.children[0].value; // the node property name is stored in its first child node
+                const imgJsonPath: string[] = Constants.imgDeploymentManifestJsonPath.slice(Constants.moduleNameDeploymentManifestJsonPathIndex + 1); // image node JSON path relative to module node
+                const imageNode: parser.Node = parser.findNodeAtLocation(moduleNode.children[1], imgJsonPath); // the node value is stored in its second child node
+
+                if (!imageNode) {
+                    continue;
+                }
+
+                const imageUrl: string = imageNode.value;
+                if (imageUrl.search(pattern) !== -1) {
```
Comment (adashen) [inline src/common/configIntelliSenseProvider.ts]: !== -1 or ===0?

File: src/extension.ts (hunk 1 above)
Comment (LazarusX) [reply src/extension.ts]: I have split it into corresponding classes.

File: src/common/configIntelliSenseProvider.ts (hunk 2 above)
Comment (LazarusX) [reply src/common/configIntelliSenseProvider.ts]: Fixed. Thanks.

File: src/common/configIntelliSenseProvider.ts (hunk 3 above)
Comment (LazarusX) [reply src/common/configIntelliSenseProvider.ts]: Fixed. Thanks.

File: src/common/configIntelliSenseProvider.ts (hunk 4 above)
Comment (LazarusX) [reply src/common/configIntelliSenseProvider.ts]: Fixed. Thanks.

**Pull request merged**